### PR TITLE
Provide a better error for setState on a defunct State

### LINKED ISSUE #1695: Confusing error: _debugLifecycleState != _StateLifecycle.defunct
In `setState` in framework.dart, there's an assert that reads:

`assert(_debugLifecycleState != _StateLifecycle.defunct);`

The meaning of this assert is difficult for a beginner to divine. This assert should be in the documentation for setState to say something like "setState() was called by a widget that was not included in the widget tree in the last call to build()."

This assert is easy to hit with an AnimationController continuing to call an old object.

### PR DESCRIPTION
**Pull request by abarth**
Fixes #1695

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -334,7 +334,16 @@ abstract class State<T extends StatefulComponent> {
   /// component will not be scheduled for rebuilding, meaning that its rendering
   /// will not be updated.
   void setState(VoidCallback fn) {
-    assert(_debugLifecycleState != _StateLifecycle.defunct);
+    assert(() {
+      if (_debugLifecycleState == _StateLifecycle.defunct) {
+        throw new WidgetError(
+          'setState called after state was disposed: $this\n'
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: `setState()`

**Comment by Hixie** [comment]
LGTM, though if @jimbeveridge has any suggestions for better wording that would have made this clearer, that's fine too.

**Comment by jimbeveridge** [comment]
How about:

'setState() called after state was disposed: $this\n'
'This happens happens if you call setState() in a widget that no\n'
'longer appears in the widget tree. See the manual for more information.'

The doc could say something to the effect of:
/// If you receive "setState() called after state was disposed,", then Flutter
/// deleted a State object because its related Component was not part of the
/// most recent call to build(). It's common for this to happen with callbacks for
/// timers or animations, which still get called even if their State object is
/// disposed. One solution is to check "mounted" in the callback to make
/// sure that the State object is still in the widget tree.

**Comment by abarth** [comment]
@jimbeveridge I've attempted to incorporate your ideas in the text.

**Comment by abarth** [comment]
I'm going to land this, but I'm open to feedback on the text.

**Comment by jimbeveridge** [comment]
lgtm.

**Pull request merged**